crates/datasources/src/object_store/mod.rs

```diff
@@ -31,35 +48,275 @@ impl FromStr for FileType {
+    {
+        let mut objects = Vec::new();
+        for location in locations {
+            let path = self.access.path(location.as_ref())?;
+            let meta = self.access.object_meta(self.store.clone(), &path).await?;
+            objects.push(meta);
         }
+        Ok(objects)
+    }
+
+    pub async fn into_table_provider(
+        self,
+        file_access: Arc<dyn FileTypeAccess>,
+        objects: Vec<ObjectMeta>,
+        predicate_pushdown: bool,
+    ) -> Result<Arc<dyn TableProvider>> {
+        let store = self.store;
+        let arrow_schema = file_access.get_schema(store.clone(), &objects).await?;
+        let base_url = self.access.base_url()?;
+
+        Ok(Arc::new(ObjStoreTableProvider {
+            store,
+            arrow_schema,
+            base_url,
+            objects,
+            file_access,
+            predicate_pushdown,
+        }))
     }
 }
 
-#[async_trait::async_trait]
-pub trait TableAccessor: Send + Sync {
-    fn store(&self) -> &Arc<dyn ObjectStore>;
-
-    fn object_meta(&self) -> &Arc<ObjectMeta>;
-    /// returns the base path.
-    /// Example:
-    /// - s3://bucket_name/path/to/file.parquet -> s3://bucket_name
-    /// - http://domain.com/path/to/file.parquet -> http://domain.com
-    fn base_path(&self) -> String;
-    /// returns the location of the file.
-    /// Example:
-    /// - s3://bucket_name/path/to/file.parquet -> path/to/file.parquet
-    fn location(&self) -> String;
-    async fn into_table_provider(self, predicate_pushdown: bool) -> Result<Arc<dyn TableProvider>>;
+pub struct ObjStoreTableProvider {
+    store: Arc<dyn ObjectStore>,
+    arrow_schema: SchemaRef,
+    base_url: ObjectStoreUrl,
+    objects: Vec<ObjectMeta>,
+    file_access: Arc<dyn FileTypeAccess>,
+    predicate_pushdown: bool,
+}
+
+#[async_trait]
+impl TableProvider for ObjStoreTableProvider {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.arrow_schema.clone()
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::View
+    }
+
+    async fn scan(
+        &self,
+        ctx: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> DatafusionResult<Arc<dyn ExecutionPlan>> {
+        let config = FileScanConfig {
+            object_store_url: self.base_url.clone(),
+            file_schema: self.arrow_schema.clone(),
+            file_groups: vec![self.objects.iter().map(|o| o.clone().into()).collect()],
+            statistics: Default::default(),
+            projection: projection.cloned(),
+            limit,
+            table_partition_cols: Vec::new(),
+            output_ordering: Vec::new(),
+            infinite_source: false,
+        };
+
+        // We register the store at scan time so that it can be used by the
+        // exec plan.
+        ctx.runtime_env()
+            .register_object_store(self.base_url.as_ref(), self.store.clone());
```

Comment:
Since we're now doing this here, I _believe_ we should be able to remove https://github.com/glaredb/glaredb/blob/904db3a14448568a41ea34c5857f7197c31bb3d2/crates/sqlexec/src/planner/dispatch.rs#L434-L437 (and the one for s3 too).

Reply:
nvm, just saw that it's been removed. Missed that on my first pass.